tutorials/haskell/gui-application/tutorial.md

```diff
@@ -0,0 +1,825 @@
+refer to:
+
+* The section in Gabriel Gonzalez's
+  [State of Haskell Ecosystem](https://github.com/Gabriel439/post-rfc/blob/master/sotu.md#standalone-gui-applications)
+
+* The [blog post](http://keera.co.uk/blog/2014/05/23/state-gui-programming-haskell/) by Keera Studios
+
+## Installation
+
+Please see [this article](https://wiki.haskell.org/Gtk2Hs/Installation) for
+Gtk2Hs installation. It has instructions for Windows, Linux, Mac, and
+FreeBSD.
+
+## First steps
+
+The calculator application has been chosen because its logic is very
+straightforward and we can focus on working with GTK+ framework without much
+distraction while keeping the tutorial reasonably practical.
+
+Let's start by importing some modules:
+
+```haskell
+import Control.Monad
+import Control.Monad.IO.Class
+import Data.IORef
+import Graphics.UI.Gtk hiding (Action, backspace)
+```
+
+As I said, GTK+ bindings are very imperative. All binding code lives in `IO`
+monad except for some cases that we will cover shortly.
+
+```haskell
+main :: IO ()
+main = do
+  void initGUI          -- (1)
+  window <- windowNew   -- (2)
+                        -- (3)
+  widgetShowAll window  -- (4)
+  mainGUI               -- (5)
+```
+
+1. The first thing we need to do in any program that uses GTK+ is to call
+   the `initGUI` function. This function allocates some resources and
+   prepares GTK+ for work, it also looks up command line arguments that are
+   relevant to GTK+, parses them, and returns all non-parsed arguments. For
+   our purposes, we don't need the command line arguments, so let's wrap it
+   with the `void`.
+
+2. Next, we need a window to build interface of our calculator inside it. To
+   create a new top-level window we use the `newWindow` action. It returns
+   opaque `Window` value that can be used to manipulate the window.
+
+3. After creation of a new window we typically want to change some
+   parameters and then render it. For now we just render the window “as is”,
+   but in the next section we will see how to customize widgets using
+   *attributes*.
+
+4. `widgetShowAll` works with any kind of widget. It performs all the
+   necessary allocations and makes widget passed to it as an argument
+   visible together with all its children widgets.
+
+5. `mainGUI` is the main loop. The loop listens to events such as button
+   click and mouse pointer movement and let appropriate handlers run.
+
+A note about threads. Make sure that all GTK actions happen on the same OS
+thread (note, this is different from lightweight Haskell threads). This only
+important when you compile with multi-threaded runtime, but who knows, maybe
+the need for concurrent execution will arise later, so my advice is to keep
+all GTK-related code in one thread and simplest way to do it is to keep
+everything is the main thread. For more information about multi-threaded
+GUIs with GTK+ see [here](http://dmwit.com/gtk2hs/).
```

Comment:
@mrkkrp 

- "This **is** only important when you compile with".
- "**the** simplest way to do it is to keep".